src/base-token.spec.ts

```diff
@@ -0,0 +1,185 @@
+import { AddressZero } from '@ethersproject/constants'
+import { ContractFactory } from '@harmony-js/contract'
+import { Harmony } from '@harmony-js/core'
+import { expect, use } from 'chai'
+import chaiAsPromised from 'chai-as-promised'
+import sinon from 'sinon'
+import { BaseToken } from './base-token'
+import { TEST_ADDRESS_1, TESTING_ABI, EMPTY_TEST_ADDRESS, TEST_ACCOUNT_2 } from './tests/constants'
+
+class TestToken extends BaseToken {}
+
+describe('Base Token Provider', () => {
+  let provider: TestToken
+  use(chaiAsPromised)
+
+  beforeEach(async () => {
+    const client = sinon.createStubInstance(Harmony)
+    client.contracts = sinon.createStubInstance(ContractFactory)
+    provider = new TestToken('', TESTING_ABI, client)
+  })
+
+  afterEach(async () => {
+    sinon.restore()
+  })
+
+  it('should be defined', () => {
+    expect(provider).to.not.be.undefined
+  })
+
+  describe('_getBalance', () => {
+    it('should get the number of tokens in the specified account in base provider', async () => {
+      const stub = sinon.stub(BaseToken.prototype, '_getBalance')
+      stub.withArgs(TEST_ADDRESS_1, 1).resolves()
+
+      expect(provider._getBalance(TEST_ADDRESS_1, 1)).to.eventually.to.exist
+      stub.restore()
+    })
+
+    it('should throw an error if address is not provided', async () => {
+      const stub = sinon.stub(BaseToken.prototype, '_getBalance')
+      stub.withArgs('', 1).onCall(0).rejects()
+
+      expect(provider._getBalance('', 1)).to.be.rejectedWith(Error)
+      stub.restore()
+    })
+
+    it('should throw an error if provided address is zero-address', async () => {
+      const stub = sinon.stub(BaseToken.prototype, '_getBalance')
+      stub.withArgs(AddressZero, 1).onCall(0).rejects()
+
+      expect(provider._getBalance(AddressZero, 1)).to.be.rejectedWith(Error)
+      stub.restore()
+    })
+  })
+
+  describe('setApprovalForAll', () => {
+    it('should grants or revokes permission to addressOperator to transfer the callers tokens, according to approved', async () => {
+      const stub = sinon.stub(BaseToken.prototype, 'setApprovalForAll')
```

Comment:
similar to the comment above, I don't think we should be stubbing the method itself. In these cases, what I think we should check is that `this.baseContract.methods.setApprovalForAll` is called with the correct params and nothing else.

Reply:
Copy that!